src/features/web-compat.js

```diff
@@ -249,39 +250,29 @@ export default class WebCompat extends ContentFeature {
                 this.onchange = null // noop
             }
         }
-        // Default subset based upon Firefox (the full list is pretty large right now and these are the common ones)
-        const defaultValidPermissionNames = [
-            'geolocation',
-            'notifications',
-            'push',
-            'persistent-storage',
-            'midi',
-            'accelerometer',
-            'ambient-light-sensor',
-            'background-sync',
-            'bluetooth',
-            'camera',
-            'clipboard',
-            'device-info',
-            'gyroscope',
-            'magnetometer',
-            'microphone',
-            'speaker'
-        ]
-        const validPermissionNames = settings.validPermissionNames || defaultValidPermissionNames
-        const returnStatus = settings.permissionResponse || 'prompt'
-        permissions.query = new Proxy((query) => {
+        permissions.query = new Proxy(async (query) => {
             this.addDebugFlag()
             if (!query) {
                 throw new TypeError("Failed to execute 'query' on 'Permissions': 1 argument required, but only 0 present.")
             }
             if (!query.name) {
                 throw new TypeError("Failed to execute 'query' on 'Permissions': Failed to read the 'name' property from 'PermissionDescriptor': Required member is undefined.")
             }
-            if (!validPermissionNames.includes(query.name)) {
+            if (!settings.supportedPermissions || !(query.name in settings.supportedPermissions)) {
                 throw new TypeError(`Failed to execute 'query' on 'Permissions': Failed to read the 'name' property from 'PermissionDescriptor': The provided value '${query.name}' is not a valid enum value of type PermissionName.`)
             }
-            return Promise.resolve(new PermissionStatus(query.name, returnStatus))
+            const permSetting = settings.supportedPermissions[query.name]
+            const returnName = permSetting.name || query.name
+            let returnStatus = settings.permissionResponse || 'prompt'
+            if (permSetting.native) {
+                try {
+                    const response = await this.messaging.request(MSG_PERMISSIONS_QUERY, query)
+                    returnStatus = response.state || 'prompt'
```

Comment:
Up to you but "state" is used already in the config to mean other enabled state of features and sub features. I would consider renaming; but also not that important.

Reply:
Ah, we'd need a change on the native side then too. If it's alright with you, let's keep it - it matches the API that way also.

Reply:
Ah yeah, leave it. It's nested deep so we can ignore :)